rules_test.go

```diff
@@ -33,27 +33,40 @@ func TestCheckUpgradeRules(t *testing.T) {
 		From    driver.Version
 		To      driver.Version
 		Allowed bool
+		Soft bool
 	}{
 		// Same version
-		{"1.2.3", "1.2.3", true},
+		{"1.2.3", "1.2.3", true, false},
 		// Different major
-		{"2.2.3", "1.2.3", false},
-		{"1.2.3", "2.2.3", false},
+		{"2.2.3", "1.2.3", false, false},
+		{"1.2.3", "2.2.3", false, false},
 		// Same major, different minor
-		{"3.2.2", "3.3.0", true},
-		{"3.2.2", "3.3.10", true},
-		{"3.3.2", "3.2.10", false},
-		{"3.3.2", "3.5.10", false},
-		{"3.2.2", "3.3.11111", true},
+		{"3.2.2", "3.3.0", true, false},
+		{"3.2.2", "3.3.10", true, false},
+		{"3.3.2", "3.2.10", false, false},
+		{"3.3.2", "3.5.10", false, false},
+		{"3.2.2", "3.3.11111", true, false},
 		// Same major & minor, different patch
-		{"3.2.2", "3.2.88", true},
-		{"3.2.88", "3.2.8", true},
-		{"3.2.88", "3.2.rc7", true},
+		{"3.2.2", "3.2.88", true, false},
+		{"3.2.88", "3.2.8", true, false},
+		{"3.2.88", "3.2.rc7", true, false},
+		// Soft
+		{"3.2.1", "3.3.1", true, true},
+		{"3.2.1", "3.4.8", true, true},
+		{"3.2.1", "3.5.rc7", true, true},
 	}
 	for _, test := range tests {
+		ur := CheckUpgradeRules
```

Comment:
Suggest  `checkUpgradeRules := CheckUpgradeRules`

more readable

Same for second function